// Copyright 2023 DeepMind Technologies Limited
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package agent;

service Agent {
  // Initialize MJPC Agent.
  rpc Init(InitRequest) returns (InitResponse);
  // Set state of the MJPC Agent.
  rpc SetState(SetStateRequest) returns (SetStateResponse);
  // Get the current action from the Agent.
  rpc GetAction(GetActionRequest) returns (GetActionResponse);
  // Compute one plan step.
  rpc PlannerStep(PlannerStepRequest) returns (PlannerStepResponse);
  // Reset the Agent.
  rpc Reset(ResetRequest) returns (ResetResponse);
  // Set a task parameter.
  rpc SetTaskParameter(SetTaskParameterRequest)
      returns (SetTaskParameterResponse);
}

message MjModel {
  optional bytes mjb = 1;
  optional string xml = 2;
}

message InitRequest {
  optional string task_id = 1;
  optional MjModel model = 2;
}
message InitResponse {}

message State {
  optional double time = 1;
  repeated double qpos = 2 [packed = true];
  repeated double qvel = 3 [packed = true];
  repeated double act = 4 [packed = true];
  repeated double mocap_pos = 5 [packed = true];
  repeated double mocap_quat = 6 [packed = true];
  repeated double userdata = 7 [packed = true];
}

message SetStateRequest {
  State state = 1;
}
message SetStateResponse {}

message GetActionRequest {
  optional float time = 1;
}

message GetActionResponse {
  repeated float action = 1 [packed = true];
}

message PlannerStepRequest {}
message PlannerStepResponse {}

message ResetRequest {}
message ResetResponse {}

message SetTaskParameterRequest {
  string name = 1;
  double value = 2;
}
message SetTaskParameterResponse {}
